src/main/java/edu/cornell/kfs/module/ar/service/impl/CuContractsGrantsInvoiceCreateDocumentServiceImpl.java

```diff
@@ -179,23 +189,253 @@ protected boolean isExpenditureSubFund(String subFundGroupCode) {
         }
         return false;
     }
+
+    private Date determineLastBilledDateByInvoicingOption(List<ContractsAndGrantsBillingAwardAccount> awardAccounts, String invoicingOptionCode, Date awardLastBilledDate) {
+        Date computedLastBilledDate = null;
+        if (StringUtils.equalsIgnoreCase(CuArConstants.AwardInvoicingOptionCodeToName.INV_ACCOUNT.getCode(), invoicingOptionCode)
+                || StringUtils.equalsIgnoreCase(CuArConstants.AwardInvoicingOptionCodeToName.INV_CONTRACT_CONTROL_ACCOUNT.getCode(), invoicingOptionCode)) {
+            ContractsAndGrantsBillingAwardAccount accountToUse = null;
+            for (ContractsAndGrantsBillingAwardAccount awardAccount : awardAccounts) {
+                if (ObjectUtils.isNotNull(accountToUse)) {
+                    if (isNotExpenditureAccount(awardAccount)
+                            && ObjectUtils.isNotNull(awardAccount.getCurrentLastBilledDate())
+                            && ObjectUtils.isNotNull(accountToUse.getCurrentLastBilledDate())
+                            && awardAccount.getCurrentLastBilledDate().after(accountToUse.getCurrentLastBilledDate())) {
+                        accountToUse = awardAccount;
```

Comment:
this for loop could be extracted to a function so you can just return the first awardAccount that meets this condition.  That way you wouldn't have to have keep track of the extra accountToUse